### DISCUSSION TITLE
Making display load faster

### DISCUSSION THREAD

**Discussion by BillyCroan** [ImageMagick/ImageMagick#5817] [General]
I'm using ImageMagick along with Yad (Yet Another Dialog) to work through a large catalog of images and categorize them.  

I have a script that calls display to open the image at a certain percent scale.  Then it opens a yad dialog if OCR wasn't able to determine everything on its own.

I've been noticing that imagemagick 'display', itself, accounts for 0.8 to 1 seconds of time from the moment it is called by my script to the moment the window is visible.

I made that part asynchronus, but it if displays after Yad, then it steals focus from Yad and that takes more user time to switch windows.

What can i do to accelerate the rendering of display?  I considered using -update.  but it doesn't seem to take interval values less than a second.  And I think copying the files around might take more time/IO as well.  Is there a way to tell an existing display window to change the file name it is displaying without killing and relaunching the window?

maybe launch display once before any images with a null image, and store its window ID and then use -window $ID to push a new image to that window background??? but would that rescale the window to show the full image if its wider?

> Nope.  No rescaling.  -window tiles the image.  bummer.  I also really like seeing the filename in the titlebar too

Is there another tool i should use for this?

**Comment by urban-warrior** [comment]
`-window $ID` works.    You can use other options to center the new image in the display window frame.  You can also send an X event to automatically resize the window to fit the size of the image.    However, the `display` program was written before the time of robust GUI interfaces and QT5.   You likely are better off using a more modern GUI framework such as GTK+.

**Comment by BillyCroan** [comment]
I tried using the -update method just set to 1 second, and I seem to be losing the -resize 33%.  i.e. the first image it displays is resized.  but images it gets after refreshing to display updated image are full size.

Is this a bug?
I would expect a display window launched with -resize 33% to always display images at 33% even if the image changes.

I guess I could work around this by writing new image files that are resized.  But they I can't zoom in later, which I sometimes need to.  Or I could but it would get pixelated.  Usually when I need to zoom in it's because I want to see detail of the jpg that is finer than the resolution of my monitor.  not just bad eyesight

@urban-warrior if I go the -window route, I could use window manager commands to resize the window, yes. but is there a way that display's left-click menu could still zoom in/out?  I think to preserve that menu I have to use -update and just hope someday it supports a shorter interval right?

**Comment by mviereck** [comment]
I am facing similar issues.
My current workaround is to use `geeqie` image viewer.
I call it with the command `geeqie -t -r --File:"IMAGENAME"`
Advantages:
 - pretty fast.
 - the window does not grab focus.
 - with the `-r` option geeqie is called remotely, so each new image appears in the same window.
 - zoom in/out, fit window etc. is no problem.
 - can additionally show a histogram.
 - several other features I don't need myself.

Disadvantages:
 - one must store the image to disk first so geeqie can read it. (Performance impact)

-----------
If I might add a feature request:
It would be nice if ImageMagick would provide a possibility to show images it currently processes, similar to `show:` or `x:`, but a bit more advanced:
- possibility to show new images in same window.
- no need for intermediate disk storage like a separate `display` or `geeqie` instance would need.
- features like auto-scaling to window size, showing a histogram, no flickering when new image is displayed.

Use case:
I am processing a bunch of images and like to show intermediate results. This can take several hours and involve hundreds or thousands of images to show. I like to see the progress and what happens to the images. This helps me to see if the image processing works as intended and which effects changed options do have.

Currently I use `geeqie` as described above for this task. But writing each image to disk first has a severe performance impact.

However, without GPU acceleration for rescaling my feature request might be less useful than I imagine yet, because of the performance impact if CPU is needed for rescaling. That would outweight the performance win of not storing to disk first. Though, if I don't use scaling, the performance win might be great.

-------------
Some outlook:

Basically, `geeqie` can do everything me and @BillyCroan  need except direct communication with `magick`. If `magick` and `geeqie` could communicate in a way that `magick` could send `mpr:` or `.mpc` directly to geeqie, this would provide best possible performance for magick and all features one can wish from an image viewer. 

I once had a friendly communication with the developer of geeqie and can imagine that he is open for some cooperation.

I am aware that the format of `.mpc` depends on the imagemagick version and also on some hardware properties.
As imagemagick and geeqie would normally run on the same machine, the hardware properties would not differ.
So if `.mpc` does not vary too much across imagemagick versions, geeqie could be able to read it if it includes some imagemagicks libs. 
And if magick would provide an option like `-show:geeqie`, magick could write a temporary `.mpc` and send its path directly to geeqie. 
Providing an `mpr:` is likely too difficult. An `.mpc` file could be stored in `XDG_RUNTIME_DIR` so it is kept in RAM and needs no harddisk i/o.
